### PR TITLE
IntersectionObserver: Intersect with Owner element

### PR DESCRIPTION
**Pull request by jridgewell**
When an element has an Owner, give the intersection accounting for the owner and viewport. Essentially, every Owner element defines it's own coordinate system so this just calculates the intersection based on that.

Fixes https://github.com/ampproject/amphtml/issues/4048.

### CODE REVIEW

**Comment by dvoytenko** [comment]
This looks good to me, but adding @cramforce for another pair of eyes.

File: src/3p-frame.js [hunk 1]
```diff
@@ -85,10 +83,7 @@ function getFrameAttributes(parentWindow, element, opt_type) {
     mode: getModeObject(),
     hidden: !viewer.isVisible(),
     amp3pSentinel: generateSentinel(parentWindow),
-    initialIntersection: getIntersectionChangeEntry(
-        Date.now(),
-        viewportFor(parentWindow).getRect(),
-        element.getLayoutBox()),
+    initialIntersection: element.getIntersectionChangeEntry(),
```
Comment (cramforce) [inline src/3p-frame.js]: Finally :)

File: test/functional/test-intersection-observer.js [hunk 2]
```diff
@@ -25,31 +25,45 @@ import * as sinon from 'sinon';
...
+      'width': 100,
+      'height': 100,
+      'bottom': 200,
+      'right': 100,
+      'x': 0,
+      'y': 100,
+    });
+    expect(change.boundingClientRect).to.deep.equal({
       'left': 50,
-      'top': -50,
```
Comment (jridgewell) [inline test/functional/test-intersection-observer.js]: Note that these three changes are actually incorrect on Prod. They now report the correct values.

File: src/intersection-observer.js [hunk 3]
```diff
@@ -14,53 +14,89 @@
...
+ *     there is an owner.
+ * @param {!./layout-rect.LayoutRectDef} viewport The viewport's layout rect.
  * @return {!IntersectionObserverEntry} A change entry.
  * @private
  */
-export function getIntersectionChangeEntry(
-    measureTime, rootBounds, elementLayoutBox) {
+export function getIntersectionChangeEntry(element, owner, viewport) {
+  dev().assert(element.width >= 0 && element.height >= 0,
+      'Negative dimensions in ad.');
```
Comment (cramforce) [inline src/intersection-observer.js]: s/ad/element

File: src/custom-element.js [hunk 4]
```diff
@@ -874,11 +874,10 @@ function createBaseAmpElementProto(win) {
...
   ElementProto.getIntersectionChangeEntry = function() {
     const box = this.implementation_.getIntersectionElementLayoutBox();
-    const rootBounds = this.implementation_.getViewport().getRect();
-    return getIntersectionChangeEntry(
-        Date.now(),
-        rootBounds,
-        box);
+    const owner = this.resources_.getResourceForElement(this).getOwner();
+    const viewportBox = this.implementation_.getViewport().getRect();
+    const ownerBox = owner && owner.getLayoutBox();
```
Comment (cramforce) [inline src/custom-element.js]: This is fine for now, but I think eventually we might need this to be recursive.

**Comment by cramforce** [comment]
LGTM

File: test/functional/test-intersection-observer.js (hunk 2 above)
Comment (cramforce) [inline test/functional/test-intersection-observer.js]: But no bugs in the `intersectionRect`?

Comment (jridgewell) [inline test/functional/test-intersection-observer.js]: Well, besides it not accounting for the owner element. But no, `intersectionRect` had the correct implementation.

**Pull request merged**